mycroft/util/process_utils.py

```diff
@@ -121,3 +121,79 @@ def echo(message):
             # Listen for messages and echo them for logging
             LOG(name).info("BUS: {}".format(message))
     return echo
+
+
+class ProcessStatus:
+    def __init__(self, name, bus,
+                 on_started=None, on_alive=None, on_ready=None,
+                 on_stopping=None, on_error=None):
+        # Messagebus connection
+        self.bus = bus
+        self.name = name
+
+        # Callback functions
+        self.on_started = on_started
+        self.on_alive = on_alive
+        self.on_ready = on_ready
+        self.on_error = on_error
+
+        # State variables
+        self.is_started = False
+        self.is_alive = False
+        self.is_ready = False
+
+        self._register_handlers()
+
+    def _register_handlers(self):
+        self.bus.on('mycroft.{}.is_alive'.format(self.name), self.check_alive)
+        self.bus.on('mycroft.{}.is_ready'.format(self.name),
+                    self.check_ready)
+        self.bus.on('mycroft.{}.all_loaded'.format(self.name),
+                    self.check_ready)
+
+    def check_alive(self, message=None):
+        """Respond to is_alive status request."""
+        if message:
+            status = {'status': self.is_alive}
+            self.bus.emit(message.response(data=status))
+        return self.is_alive
+
+    def check_ready(self, message=None):
+        """ Respond to all_loaded status request."""
+        print("CHECKING READY!")
+        if message:
+            status = {'status': self.is_ready}
+            self.bus.emit(message.response(data=status))
+
+        return self.is_ready
+
+    def set_started(self):
+        """Process is started."""
+        self.is_started = True
+        if self.on_started:
+            self.on_started()
+
+    def set_alive(self):
+        """Basic loading is done."""
+        self.is_alive = True
+        self.is_started = True
+        if self.on_alive:
+            self.on_alive()
+
+    def set_ready(self):
+        """All loading is done."""
+        self.is_alive = True
+        self.is_ready = True
+        if self.on_ready:
+            self.on_ready()
+
+    def set_stopping(self):
+        self.is_ready = False
+        if self.on_stopping:
+            self.on_stopping()
+
+    def set_error(self, err=''):
+        self.is_ready = False
+        self.is_alive = False
+        if self.on_stopping:
+            self.on_ready(err)
```

Comment:
@forslund should this be
```
if self.on_error:
    self.on_error(err)
```
or am I missing something?

Reply:
Yes indeed, I have that changed locally but not committed apparently along with a couple other minor things...

Reply:
Here's my local changes + some improved docstrings: https://github.com/forslund/mycroft-core/commit/8be5fa43516a4093ae6fd36f4f0904dcd0d5c56c